File: api/oauth.go
package api

// OAuthToken is the API message for OAuthToken.
type OAuthToken struct {
	AccessToken  string `jsonapi:"attr,accessToken" `
	RefreshToken string `jsonapi:"attr,refreshToken"`
	ExpiresTs    int64  `jsonapi:"attr,expiresAt"`
}

// OAuthConfig is the API message for OAuthConfig.
type OAuthConfig struct {
	Endpoint      string `jsonapi:"attr,endpoint"`
	ApplicationID string `jsonapi:"attr,applicationId"`
	Secret        string `jsonapi:"attr,secret"`
	RedirectURL   string `jsonapi:"attr,redirectUrl"`
}

// ExchangeToken is the API message for exchanging token.
type ExchangeToken struct {
	ID     int         `jsonapi:"attr,vcsId"`
	Code   string      `jsonapi:"attr,code"`
	Config OAuthConfig `jsonapi:"relation,config"`
}

File: frontend/src/store/pinia-modules/oauth.ts
import axios from "axios";
import { VCSId, OAuthToken, ResourceObject } from "@/types";
import { defineStore } from "pinia";

const convert = (raw: ResourceObject): OAuthToken => {
  const attr = raw.attributes;
  return {
    accessToken: attr.accessToken as string,
    refreshToken: attr.refreshToken as string,
    expiresTs: attr.expiresTs as number,
  };
};

export const useOAuthStore = defineStore("oauth", {
  actions: {
    async exchangeVCSToken({ vcsId, code }: { vcsId: VCSId; code: string }) {
      const data = (
        await axios.post(`/api/oauth/vcs/${vcsId}/exchange-token`, null, {
          headers: { code },
        })
      ).data.data;

      const token = convert(data);
      return token;
    },
  },
});
